tests/unit/conftest.py

```diff
@@ -99,28 +99,29 @@ def engine(pytestconfig):
 @pytest.fixture
 def connection(engine):
     """Return a connection to the database that rolls back automatically."""
-    with engine.begin() as conn:
-        savepoint = conn.begin_nested()
-        yield conn
-        savepoint.rollback()
+    conn = engine.connect()
+    yield conn
+    conn.close()
 
 
 @pytest.fixture
 def dbsession(connection):
     """Return a database session that rolls back."""
-    test_sessionmaker = sessionmaker(autocommit=False, autoflush=False, bind=connection)
-    db = test_sessionmaker()
+    transaction = connection.begin()
+    session = Session(autocommit=False, autoflush=False, bind=connection)
+    nested = connection.begin_nested()
 
-    def test_get_db():
-        db.begin_nested()
-        try:
-            yield db
-        finally:
-            db.close()
+    # If the application code calls session.commit, it will end the nested
+    # transaction. Need to start a new one when that happens.
+    @event.listens_for(session, "after_transaction_end")
+    def end_savepoint(session, transaction):
+        nonlocal nested
+        if not nested.is_active:
+            nested = connection.begin_nested()
```

Comment:
Did you assemble these lines yourself or is there a link to docs?

Reply:
I included a link in the commit message, but yeah, probably better to include the direct link to the docs here as a comment